db/migrate/20210414203512_convert_primary_language_to_iso_code.rb

```diff
@@ -0,0 +1,124 @@
+class ConvertPrimaryLanguageToIsoCode < ActiveRecord::Migration[6.1]
+  def match_language(lang)
+    return nil if lang.nil?
+
+    # spa-PR is the only iso-code that involves case. it will not be properly matched if
+    # we call downcase on the input
+    lang = lang.casecmp('spa-pr')&.zero? ? 'spa-PR' : lang.to_s.downcase
+    matched_language = nil
+    matched_language = lang.to_sym if PATIENT_HELPER_FILES[:languages][lang.to_sym]
+    return matched_language unless matched_language.nil?
+
+    matched_language = PATIENT_HELPER_FILES[:languages].find { |_key, val| val[:display]&.casecmp(lang)&.zero? }
+    return matched_language[0] unless matched_language.nil?
+
+    matched_language = PATIENT_HELPER_FILES[:languages].find { |_key, val| val[:iso6391code]&.casecmp(lang)&.zero? }
+    return matched_language[0] unless matched_language.nil?
+
+    matched_language
+  end
+
+  def up
+    add_column :patients, :legacy_primary_language, :string
+    add_column :patients, :legacy_secondary_language, :string
+    execute <<-SQL.squish
+      UPDATE patients
+      SET legacy_primary_language = primary_language, legacy_secondary_language = secondary_language
+      WHERE purged = FALSE
+    SQL
+
+    # automatic_translations will take the format of { 'English' : 'eng' }
+    # Where `English` is plucked directly from the DB
+    automatic_translations = {}
+
+    # custom_translations will take the format of { 'portugese' : 'eng' }
+    # Where `portugese` is manually defined as a value that was manually selected from the DB (from manual_translations)
+    # The only different between `custom_translations` and `automatic_translations` is that
+    # when languages are updated with custom_translations, the Patients will have a HistoryItem
+    # added to them. Patients who are translated with automatic_translations will NOT have that item on them
+    custom_translations = {}
+
+    # Update this hash as necessary with any manual translations/misspellings
+    # The 'key' part can be display name, or iso code. It will be correctly translated before any execution happens
+    manual_translations = {
+      'bilingual english/spanish': 'english',
+      'portugese': 'portuguese',
+      'bos': 'bosnian',
+      'laotian': 'lao',
+      'pelugu': 'telugu',
+      'telegu': 'telugu',
+      'philippines': 'filipino',
+      'punjabi': 'panjabi',
+      'frysian': 'frisian',
+      'haitian creole': 'haitian',
+    }
+    # Get one master hash of languages to iso codes
+    manual_translations.each do |key, value|
+      custom_translations[key.to_sym] = (match_language(value) || 'eng').to_s
+    end
+
+    existing_primary_languages = Patient.where(purged: false).distinct.pluck(:primary_language)
```

Comment:
there seems to be a bit of repetition for operations performed on primary and secondary languages which I think can be refactored out into a separate method

Reply:
I noticed the same thing, but figured it was best for readability to keep them separate (since there's not a lot of code). I'll look into combining stuff.

Reply:
one way we can sometimes make it more readable is by writing it like:
```
Patient.where(purged: false)
       .where.not(primary_language: nil)
       .distinct
       .pluck(:primary_language)
```

Reply:
fixed!